src/proto/linear_map.rs

```diff
@@ -0,0 +1,611 @@
+    /// allocator may still inform the map that there is more space than
+    /// necessary. Therefore capacity cannot be relid upon to be minimal.
+    pub fn shrink_to_fit(&mut self) {
+        self.storage.shrink_to_fit();
+    }
+
+    /// Returns the number of elements in the map.
+    pub fn len(&self) -> uint {
+        self.storage.len()
+    }
+
+    /// Returns true if the map contains no elements.
+    pub fn is_empty(&self) -> bool {
+        self.storage.is_empty()
+    }
+
+    /// Clears the map, removing all elements. Keeps the allocated memory for
+    /// reuse.
+    pub fn clear(&mut self) {
+        self.storage.clear();
+    }
+
+    /// An iterator visiting all key-value pairs in arbitrary order. Iterator
+    /// element type is `(&'a K, &'a V)`.
+    pub fn iter<'a>(&'a self) -> Iter<'a, K, V> {
+        fn ref_<A,B>(&(ref v1, ref v2): &(A, B)) -> (&A, &B) { (v1, v2) }
+        Iter { iter: self.storage.iter().map(ref_ as fn(&'a (K, V)) -> (&'a K, &'a V)) }
+    }
+
+    /// An iterator visiting all key-value pairs in arbitrary order with
+    /// mutable references to the values. Iterator element type is `(&'a K, &'a
+    /// mut V)`.
+    pub fn iter_mut<'a>(&'a mut self) -> IterMut<'a, K, V> {
+        fn ref_<A,B>(&(ref v1, ref mut v2): &mut (A, B)) -> (&A, &mut B) { (v1, v2) }
+        IterMut { iter: self.storage.iter_mut().map(ref_ as fn(&'a mut (K, V)) -> (&'a K, &'a mut V)) }
+    }
+
+    /// An iterator visiting all keys in arbitrary order. Iterator element type
+    /// is `&'a K`.
+    pub fn keys<'a>(&'a self) -> Keys<'a, K, V> {
+        fn first<A,B>((v, _): (A, B)) -> A { v }
+        Keys { iter: self.iter().map(first as fn((&'a K, &'a V)) -> &'a K) }
+    }
+
+    /// An iterator visiting all values in arbitrary order. Iterator element
+    /// type is `&'a V`.
+    pub fn values<'a>(&'a self) -> Values<'a, K, V> {
+        fn second<A,B>((_, v): (A, B)) -> B { v }
+        Values { iter: self.iter().map(second as fn((&'a K, &'a V)) -> &'a V) }
+    }
+
+    /// Returns a reference to the value corresponding to the key.
+    pub fn get<'a>(&'a self, key: &K) -> Option<&'a V> {
+        for (k, v) in self.iter() {
+            if key == k {
+                return Some(v);
+            }
+        }
+        None
+    }
+
+    /// Returns a mutable reference to the value corresponding to the key.
+    pub fn get_mut<'a>(&'a mut self, key: &K) -> Option<&'a mut V> {
+        for (k, v) in self.iter_mut() {
+            if key == k {
+                return Some(v);
+            }
+        }
+        None
+    }
+
+    /// Returns true if the map contains a value to the specified key.
+    pub fn contains_key(&self, key: &K) -> bool {
+        self.get(key).is_some()
+    }
+
+    /// Inserts a key-value pair into the map. If the key already had a value
+    /// present in the map, it is returned. Otherwise, `None` is returned.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
```

Comment:
Doesn't pass the borrow checker, because `self.get_mut` borrows `self` and the type checker isn't able to see that there's no actual borrow in the `None` case. I can still derive an enhancement to that function though.